### DISCUSSION TITLE
JNI Error Handling

### DISCUSSION THREAD

**Discussion by iprovalo** [k2-fsa/sherpa-onnx#1426] [General]
I am curious if it would be beneficial to add some exception handling at the JNI layer in order to avoid the app crashes.

For example, in the offline-recognizer.cc:

```
SHERPA_ONNX_EXTERN_C
JNIEXPORT void JNICALL Java_com_k2fsa_sherpa_onnx_OfflineRecognizer_decode(
    JNIEnv *env, jobject /*obj*/, jlong ptr, jlong streamPtr) {
  auto recognizer = reinterpret_cast<sherpa_onnx::OfflineRecognizer *>(ptr);
  auto stream = reinterpret_cast<sherpa_onnx::OfflineStream *>(streamPtr);

  try {
    recognizer->DecodeStream(stream);
  } catch (const std::exception& e) {
      jclass exClass = env->FindClass("java/lang/RuntimeException");
      if (exClass != nullptr) {
          env->ThrowNew(exClass, e.what());
      }
  }
} 

```

I am observing some crashes like this on Android:

```
  #00  pc 0x0000000000b5def4  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)
  #01  pc 0x0000000000b49894  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)
  #02  pc 0x0000000000b7599c  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)
  #03  pc 0x0000000000b75558  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)
  #04  pc 0x0000000000b76b78  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)
  #05  pc 0x0000000000459c04  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)
  #06  pc 0x000000000045b67c  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)
  #07  pc 0x000000000042eb9c  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)
  #08  pc 0x00000000000e9b60  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libsherpa-onnx-jni.so (BuildId: c816740f0b120054222ce04e10479ec3252b4474)
  #09  pc 0x00000000000e9a94  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libsherpa-onnx-jni.so (BuildId: c816740f0b120054222ce04e10479ec3252b4474)
  #10  pc 0x00000000000fa7cc  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libsherpa-onnx-jni.so (BuildId: c816740f0b120054222ce04e10479ec3252b4474)
  #11  pc 0x00000000000da578  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libsherpa-onnx-jni.so (BuildId: c816740f0b120054222ce04e10479ec3252b4474)
  #12  pc 0x000000000008a020  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libsherpa-onnx-jni.so (Java_com_k2fsa_sherpa_onnx_OfflineRecognizer_decode+52) (BuildId: c816740f0b120054222ce04e10479ec3252b4474)
  #13  pc 0x00000000003343c0  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (art_jni_trampoline+112)
  #14  pc 0x000000000077fc80  /apex/com.android.art/lib64/libart.so (nterp_helper+5648)
```

**Comment by csukuangfj** [comment]
In general, `recognizer.decode()` won't throw unless
 - 1. Your input audio is super short, e.g., less than 0.1 seconds
 - 2. Or your input audio is super long, e.g., several minutes long.

Could you describe how you get the above error?

**Reply by iprovalo** [reply]
Thank you, @csukuangfj !

 It happens in runSecondPass() when decode() is called:

```
...
   private val sampleRateInHz = 16000
   private val channelConfig = AudioFormat.CHANNEL_IN_MONO
   private val audioFormat = AudioFormat.ENCODING_PCM_16BIT
...

 private fun processSamples(): String {
        Log.i(TAG, "processing samples")

        val minBufferSize = AudioRecord.getMinBufferSize(sampleRateInHz, channelConfig, audioFormat)
        val bufferSize = if (minBufferSize > 0) minBufferSize * 2 else 512

        val buffer = ShortArray(bufferSize)

        while (true) {
            val ret = audioRecord?.read(buffer, 0, buffer.size)
            if (ret != null && ret > 0) {
                val samples = FloatArray(ret) { buffer[it] / 32768.0f }

                vad.acceptWaveform(samples)
                while(!vad.empty()) {
                    var segment = vad.front()
                    val text = runSecondPass(segment.samples)

                    vad.pop();

                    if (text.isNotBlank()) {
                        return text
                    }


                }
            }
        }
    }


...

    private fun runSecondPass(samples: FloatArray): String {
        val stream = offlineRecognizer.createStream()
        stream.acceptWaveform(samples, sampleRateInHz)
        offlineRecognizer.decode(stream)
        val result = offlineRecognizer.getResult(stream)
        stream.release()
        return result.text
    }
```

**Reply by csukuangfj** [reply]
> In general,  `recognizer.decode()` won't throw unless
>  - 1. Your input audio is super short, e.g., less than 0.1 seconds
>  - 2. Or your input audio is super long, e.g., several minutes long.
> 
> Could you describe how you get the above error?

how long is the audio for.the 2nd pass?

**Reply by iprovalo** [reply]
@csukuangfj thank you, I have not reproduced this one on my device yet.  These happen on the field devices.  I will try creating the condition.

Would it be a good idea to introduce the JNI layer exception handling, for example, in order to handle the unexpected exceptions?

**Reply by csukuangfj** [reply]
If you throw exceptions from C++, then you must handle them in the java/kotlin code.

Could you set the result to empty in C++ instead of throwing an exception?

**Reply by iprovalo** [reply]
@csukuangfj that sounds reasonable, thank you!

**Reply by iprovalo** [reply]
@csukuangfj do you mean something like this:

```
SHERPA_ONNX_EXTERN_C
JNIEXPORT void JNICALL Java_com_k2fsa_sherpa_onnx_OfflineRecognizer_decode(
    JNIEnv * /*env*/, jobject /*obj*/, jlong ptr, jlong streamPtr) {
  auto recognizer = reinterpret_cast<sherpa_onnx::OfflineRecognizer *>(ptr);
  auto stream = reinterpret_cast<sherpa_onnx::OfflineStream *>(streamPtr);

  try {
    recognizer->DecodeStream(stream);
  } catch (const std::exception& e) {
    SHERPA_ONNX_LOGE("Error while decoding stream."+ e.what());

    sherpa_onnx::OfflineRecognitionResult defaultResult;
    defaultResult.text = "";
    defaultResult.tokens = {""};
    defaultResult.timestamps = {0.0f};
    defaultResult.lang = "";
    defaultResult.emotion = "";
    defaultResult.event = "";
    stream->SetResult(defaultResult);
  }
}
```

**Reply by csukuangfj** [reply]
you can return directly from the catch without setting anything.

**Reply by iprovalo** [reply]
What about `Java_com_k2fsa_sherpa_onnx_OfflineRecognizer_getResult`()?  Kotlin/Java will call that next and expect the result to be available on the stream object.

**Reply by iprovalo** [reply]
Honestly, I feel it is better to throw a Java `RuntimeException` from `Java_com_k2fsa_sherpa_onnx_OfflineRecognizer_decode`, and let Java/Kotlin handle it.  Because the app UI may choose to display something different to the user in these cases.  

I have a couple of crashes like this still happening in the field.  The other ones are in the Vad APIs.

**Reply by iprovalo** [reply]
Here is a proposed PR: https://github.com/k2-fsa/sherpa-onnx/pull/1452

**Reply by csukuangfj** [reply]
> Honestly, I feel it is better to throw a Java `RuntimeException` from `Java_com_k2fsa_sherpa_onnx_OfflineRecognizer_decode`, and let Java/Kotlin handle it.  Because the app UI may choose to display something different to the user in these cases.  
> 
> I have a couple of crashes like this still happening in the field.  The other ones are in the Vad APIs.

could you post the audio duration when iy throws?

**Reply by iprovalo** [reply]
@csukuangfj these crashes happen on field devices, I was not able to reproduce them.  I only get abbreviated crash reports.  Unfortunately, it doesn't have the duration data.

**Reply by csukuangfj** [reply]
it usually indicates that the input data is too short, e.g. less than 0.2 seconds.

**Reply by iprovalo** [reply]
@csukuangfj I am more concerned about any unexpected errors.

**Reply by iprovalo** [reply]
@csukuangfj if I want to add the debug symbols to my Android build, do I just need to modify this `-DCMAKE_BUILD_TYPE=RelWithDebInfo` flag in the `build*.sh`, correct? 

```
cmake -DCMAKE_TOOLCHAIN_FILE="$ANDROID_NDK/build/cmake/android.toolchain.cmake" \
    -DSHERPA_ONNX_ENABLE_TTS=$SHERPA_ONNX_ENABLE_TTS \
    -DSHERPA_ONNX_ENABLE_SPEAKER_DIARIZATION=$SHERPA_ONNX_ENABLE_SPEAKER_DIARIZATION \
    -DSHERPA_ONNX_ENABLE_BINARY=$SHERPA_ONNX_ENABLE_BINARY \
    -DBUILD_PIPER_PHONMIZE_EXE=OFF \
    -DBUILD_PIPER_PHONMIZE_TESTS=OFF \
    -DBUILD_ESPEAK_NG_EXE=OFF \
    -DBUILD_ESPEAK_NG_TESTS=OFF \
```
```
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
```
```
    -DBUILD_SHARED_LIBS=ON \
    -DSHERPA_ONNX_ENABLE_PYTHON=OFF \
    -DSHERPA_ONNX_ENABLE_TESTS=OFF \
    -DSHERPA_ONNX_ENABLE_CHECK=OFF \
    -DSHERPA_ONNX_ENABLE_PORTAUDIO=OFF \
    -DSHERPA_ONNX_ENABLE_JNI=$SHERPA_ONNX_ENABLE_JNI \
    -DSHERPA_ONNX_ENABLE_C_API=$SHERPA_ONNX_ENABLE_C_API \
    -DCMAKE_INSTALL_PREFIX=./install \
    -DANDROID_ABI="arm64-v8a" \
    -DANDROID_PLATFORM=android-21 ..
```

**Reply by iprovalo** [reply]
I have two API's right now frequently throwing a similar exception where the crash stack ends with some `onnxruntime`: 

`  #00  pc 0x0000000000b5def4  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)`

It's `Java_com_k2fsa_sherpa_onnx_OfflineRecognizer_decode` and `Java_com_k2fsa_sherpa_onnx_Vad_acceptWaveform`.

Another frequent crash is with `Java_com_k2fsa_sherpa_onnx_Vad_reset`.  This one is not related to `onnxruntime`.

**Reply by csukuangfj** [reply]
> I have two API's right now frequently throwing a similar exception where the crash stack ends with some `onnxruntime`: 
> 
> `  #00  pc 0x0000000000b5def4  /data/app/~~er-YO3C0U2mhaw4uNlItFg==/com.___.android-5IuAA6NTrxfLlXlcKIMYtQ==/split_config.arm64_v8a.apk!libonnxruntime.so (BuildId: 96aa9c7a798685f988b107ba72fae8918594328d)`
> 
> It's `Java_com_k2fsa_sherpa_onnx_OfflineRecognizer_decode` and `Java_com_k2fsa_sherpa_onnx_Vad_acceptWaveform`.
> 
> Another frequent crash is with `Java_com_k2fsa_sherpa_onnx_Vad_reset`.  This one is not related to `onnxruntime`.

By the way, are you using multithreads?

Vad and Stream are not thread safe.

**Reply by iprovalo** [reply]
It's a kotlin background IO thread which invokes the listening method.  UI stays locked for any further actions.  I keep the vad and recognizer within a synchronized method.  Technically, there should not be any two threads accessing the APIs in my case.  I will double check all of these objects again.

**Reply by csukuangfj** [reply]
Also, please don't forget that the  stream object is also not thread safe.

**Reply by iprovalo** [reply]
Thank you!  I will double-check everything.  

What about adding the native debug symbols file for more detailed crash report?  Do you have the instructions for for doing that by chance?

**Reply by csukuangfj** [reply]
> Thank you! I will double-check everything.
> 
> What about adding the native debug symbols file for more detailed crash report? Do you have the instructions for for doing that by chance?

Sorry, I don't have much experience in debugging C++ code with Android.

**Comment by iprovalo** [comment]
@csukuangfj I was able to create the debug symbols file for sherpa-onnx project, but I am missing the onnxruntime debug symbols library.  Can you please point me to the native build for onnxruntime, so I can create one as well?

**Reply by csukuangfj** [reply]
sorry，i don't know about it.

By the way, if you need to debug onnxruntime， it indicates there are issues about your code.

**Reply by iprovalo** [reply]
@csukuangfj it is most likely some problem with my code, but this issue is happening on a small percentage of a wide range of devices in the field.  It is a condition which is hard to debug and I have not been able to reproduce it on any of my devices.  Adding the symbolized non-stripped builds is one way to get to the bottom of this.

This is what I was looking for: 
https://github.com/csukuangfj/onnxruntime-libs/blob/master/.github/workflows/aarch64-shared.yaml#L28

**Reply by csukuangfj** [reply]
it is not for Android.

We use the pre built onnxruntime lib for Android from Microsoft.
